File: src/library/blas.f90
subroutine matvec_coo(a, row, col, x, y, nnz, n)
    implicit none
    integer, intent(in) :: nnz, n
    integer, dimension(nnz), intent(in) :: row, col
    real(kind=8), dimension(nnz), intent(in) :: a
    real(kind=8), dimension(n), intent(in) :: x
    real(kind=8), dimension(n), intent(inout) :: y
    integer :: i, irow, icol
    y = 0.d0
    do i = 1, nnz
        irow = row(i) + 1
        icol = col(i) + 1
        y(irow) = y(irow) + a(i) * x(icol)
    enddo
end subroutine matvec_coo

File: src/pyfsolver/numpy/blas.py
import numpy as np

from pyfsolver.coo import COO


def matvec_coo(a: COO, x: np.ndarray, y: np.ndarray | None = None) -> np.ndarray:
    if y is None:
        y = np.zeros_like(x)
    else:
        y[:] = 0.0e0

    for i in range(a.nnz):
        irow, icol = a.row[i], a.col[i]
        y[irow] += a.data[i] * x[icol]

    return y
